bundles/org.openhab.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/builder/BridgeBuilder.java

```diff
@@ -101,4 +102,13 @@ public BridgeBuilder withLocation(@Nullable String location) {
         return (BridgeBuilder) super.withLocation(location);
     }
 
+    public BridgeBuilder withThings(@Nullable List<Thing> things) {
+        BridgeImpl bridge = ((BridgeImpl) thing);
```

Comment:
Bear with me - I am short of time. I think, the only issue left open is the removal of channels during `editThing`, that are not defined in XML and therefore must not be removed. I have no clear idea yet, how to solve this.

Reply:
It might help to add a field to the `Channel` class for tracking the origin. Then it can be removed if it was not created from the ThingType definition.

There are also some other border cases:
* The new ThingType defines a channel that was created by the user. In that case the user channel and configuration might get lost because it is replaced with that of ThingType.
* An existing channel is renamed. If the channel has configuration parameters its values will be lost when the old channel is removed.

I'm fine with skipping those border cases for now because I think they won't occur often and the changes in this PR would do more good than harm. :-)

Reply:
> It might help to add a field to the Channel class for tracking the origin.

Yes, that is true. But I am afraid it is not possible until we found a solution for #731.

btw: I prepared a new rebase for these changes including some unit tests. https://github.com/openhab/openhab-core/compare/master...cweitkamp:pr-1027